File: backend/app/services/candle_service.py
from sqlalchemy.orm import Session

from app.core.providers import (
    get_market_data_provider,
)
from app.services.market_data_service import (
    MarketDataService,
)


class CandleService:
    """
    Read-only candle service.

    Delegates market data lifecycle
    to MarketDataService.
    """

    def __init__(
        self,
        db: Session,
    ) -> None:

        provider = get_market_data_provider()

        self.market_data_service = (
            MarketDataService(
                db=db,
                provider=provider,
            )
        )

    def get_candles(
        self,
        symbol: str,
        timeframe: str,
        limit: int = 1000,
    ):

        return self.market_data_service.ensure_history(
            symbol=symbol,
            timeframe=timeframe,
            limit=limit,
        )

File: frontend/src/services/candle.service.ts
import api from "./api";
import type { Candle } from "@/types/candle";
import type { TimeFrame } from "@/types/timeframe";

export async function getCandles(
    symbol: string,
    timeframe: TimeFrame,
) {
    const response =
        await api.get<Candle[]>(
            "/candles",
            {
                params: {
                    symbol,
                    timeframe,
                    limit: 500,
                },
            },
        );

    return response.data;
}